src/lex/mod.rs

```diff
@@ -118,14 +118,11 @@ impl Lexer {
         self.chars[self.location.offset as usize..].chars()
     }
 
-    fn slice(&self, span_start: u32) -> ArcStr {
+    fn slice(&self, span_start: u32) -> Substr {
         use std::ops::Range;
 
-        let chars = self.chars.clone();
-        let chars = chars
-            .get::<Range<usize>>(self.span(span_start).span.into())
-            .unwrap_or("");
-        ArcStr::from(chars)
+        let range: Range<usize> = self.span(span_start).span.into();
+        ArcStr::from(self.chars.clone()).substr(range)
```

Comment:
This is still missing the point. ArcStr::from copies the input. You should make `self.chars` an ArcStr itself so that you only call `from` once in the whole program.

Reply:
Hmm... Actually I think it might not copy if you're creating it from an Rc. Can you use `Rc::clone` here to make sure that's the generic impl that's chosen?

Reply:
Nah, that's actually a clone of the chars.

Reply:
You were correct here, that it will copy even if creating from an Rc. We don't reuse the stdlib Arc/Rc impl for a lot of reasons: FFI safety, avoiding needing to worry about mutability or Weak, etc, ensuring our ArcStr is both a single pointer without adding an extra indirection, transparent/cheap support for literals, ... Honestly probably more.